### PR TITLE
Deprecate QuantumInstance and Opflow

### PR DESCRIPTION
**Pull request by manoelmarques**
### Details and comments
These classes/functions outside the opflow module still use opflow as method argument or internal code:

- qiskit.algorithms.time_evolvers.TimeEvolutionProblem -> argument + instance check
- qiskit.primitives.BackendEstimator -> argument
- qiskit.primitives.Estimator -> argument + instance check
- qiskit.primitives.utils.init_observable ->argument + instance check
- qiskit.primitives.base.BaseEstimator -> argument
- qiskit.result.sampled.expval.sampled_expectation_value -> argument + instance check
- ~~qiskit.synthesis.evolution.QDrift~~ -> removed in this PR

### CODE REVIEW

**Review by jakelishman — CHANGES_REQUESTED** [review]
I'll be glad to see the back of `QuantumInstance` in particular.  This is going to be one of the larger deprecations and transitions we have to navigate our users through, and I don't think we're offering enough support in this PR.

Some questions about how this deprecation is being done:

1. There are a lot of `opflow` imports remaining in `qiskit.algorithms`.  What is happening to these?  They should be migrated _before_ the deprecation.
2. There are a lot of `QuantumInstance`-related methods and properties in `algorithms`.  What is happening to these?  Should these be part of the same deprecation?
3. Removing `opflow` and `QuantumInstance` is a massive change to the algorithms/applications interface.  What transition guides do we have in place to manage this change?  We should be promoting these _heavily_ before any deprecations actually land in release Terra.

This is going to be a massive inconvenience for users, and we really need to make sure we're probably preparing them for the transition.  All the warnings we're emitting need to be clear about what the alternative is, and give them links to how they can transition their code.  We need those transition guides in place and being widely publicised, along with the reasons why, before we drop any warning messages on users.

General comments on the implementation here:
- putting "Deprecated" into a docstring isn't the proper way to show deprecations - that's the `.. deprecated::` directive.
- documentation needs updating to show the new paths, not just to `warnings.filterwarnings` the messages away
- It's not generally good to use `catch_warnings` in library code; it's slow, clunky, and often hides true warnings.  Instead, the `stacklevel` arguments should be set correctly such that Qiskit library code should be blamed for the warning where appropriate, which prevents `DeprecationWarning` from being shown to users with default filters.
- in all `deprecate_function` calls, we should be pointing users to what they should be using instead, preferably via a complete migration guide.
- the testing of the warnings here is not sufficient.  We cannot suppress warnings generated by `algorithms` and other Terra places because of `opflow` use - we need to know about them, and fix them before the warnings are emitted.  If we can't do that, how are we expecting users to do it?  If there are undeprecated public entry points in `algorithms` and elsewhere that _should_ warn on certain inputs, then they should have tests asserting that they do.


Other places to consider:
- `qiskit.utils.backend_utils` feels very `QuantumInstance`-y.  Can we deprecate that now?
- `qiskit.utils.mitigation.fitters` refers to `QuantumInstace`.  Do we need to do something about this?

File: qiskit/algorithms/evolvers/trotterization/trotter_qrte.py
Comment (manoelmarques) [reply qiskit/algorithms/evolvers/trotterization/trotter_qrte.py:69]: This PR doesn't address the modifications in all the rest of the code to remove opflow and QuantumInstance. This should be done in one or more separate PRs since, as you noted, is massive. I will change this PR to draft for now.

File: test/python/algorithms/test_shor.py
Comment (manoelmarques) [reply test/python/algorithms/test_shor.py]: Changed the statements

File: qiskit/opflow/__init__.py
Comment (manoelmarques) [reply qiskit/opflow/__init__.py]: Added statement

File: qiskit/opflow/__init__.py
Comment (manoelmarques) [reply qiskit/opflow/__init__.py]: Removed

File: qiskit/algorithms/linear_solvers/observables/absolute_average.py
Comment (manoelmarques) [reply qiskit/algorithms/linear_solvers/observables/absolute_average.py]: The class AbsoluteAverage itself is deprecated so it should keep the opflow logic. The warning filter was removed from the code sample above in order to preserve AbsoluteAverage usage.

File: qiskit/algorithms/linear_solvers/observables/matrix_functional.py
Comment (manoelmarques) [reply qiskit/algorithms/linear_solvers/observables/matrix_functional.py]: The class MatrixFunctional itself is deprecated so it should keep the opflow logic. The warning filter was removed from the code sample above in order to preserve MatrixFunctional usage.

File: qiskit/test/base.py
Comment (manoelmarques) [reply qiskit/test/base.py]: Removed filters from base.py

File: releasenotes/notes/deprecate-opflow-qi-32f7e27884deea3f.yaml
Comment (manoelmarques) [reply releasenotes/notes/deprecate-opflow-qi-32f7e27884deea3f.yaml]: Links to migration guides added to release note

File: qiskit/utils/deprecation.py
Comment (ElePT) [reply qiskit/utils/deprecation.py:20]: Got it, I'm waiting for #9676, I was just making sure in the meantime that there were no additional merge conflicts in this PR.

File: qiskit/opflow/__init__.py
Comment (ElePT) [inline qiskit/opflow/__init__.py:25]: Adding a deprecation warning at the top should be enough, right? I considered reiterating this for each submodule listed.

File: qiskit/opflow/converters/abelian_grouper.py
Comment (ElePT) [inline qiskit/opflow/converters/abelian_grouper.py:47]: Not sure about the "deprecation effort" part of the message. I now realize I should probably also split this message along 2 lines, the string is too long.

File: qiskit/utils/backend_utils.py
Comment (ElePT) [inline qiskit/utils/backend_utils.py:120]: I will go through all of the QI-related messages tomorrow, because I see that they are not consistent.

File: qiskit/utils/measurement_error_mitigation.py
Comment (ElePT) [inline qiskit/utils/measurement_error_mitigation.py:41]:
```suggestion
    Retrieve the measured qubits from transpiled circuits.
```

File: qiskit/utils/measurement_error_mitigation.py
Comment (ElePT) [inline qiskit/utils/measurement_error_mitigation.py:90]:
```suggestion
    Retrieve the measured qubits from transpiled circuits.
```

File: qiskit/utils/measurement_error_mitigation.py
Comment (ElePT) [inline qiskit/utils/measurement_error_mitigation.py:144]:
```suggestion
    """Build measurement error mitigation circuits
```

File: qiskit/utils/mitigation/_filters.py
Comment (ElePT) [inline qiskit/utils/mitigation/_filters.py:39]:
```suggestion
    Measurement error mitigation filter.
```

File: qiskit/utils/mitigation/_filters.py
Comment (ElePT) [inline qiskit/utils/mitigation/_filters.py:222]:
```suggestion
    Tensored measurement error mitigation filter.
```

File: qiskit/utils/mitigation/circuits.py
Comment (ElePT) [inline qiskit/utils/mitigation/circuits.py:30]:
```suggestion
    """Return ordered count keys.
```

File: qiskit/utils/mitigation/circuits.py
Comment (ElePT) [inline qiskit/utils/mitigation/circuits.py:54]:
```suggestion
    Return a list of measurement calibration circuits for the full
```

File: qiskit/utils/mitigation/circuits.py
Comment (ElePT) [inline qiskit/utils/mitigation/circuits.py:135]:
```suggestion
    Return a list of calibration circuits
```

File: qiskit/utils/mitigation/fitters.py
Comment (ElePT) [inline qiskit/utils/mitigation/fitters.py:219]:
```suggestion
    Measurement correction fitter for a tensored calibration.
```

File: qiskit/utils/quantum_instance.py
Comment (ElePT) [inline qiskit/utils/quantum_instance.py:128]:
```suggestion
    """Quantum Backend including execution setting."""
```

File: qiskit/utils/run_circuits.py
Comment (ElePT) [inline qiskit/utils/run_circuits.py:50]:
```suggestion
    """Find the registers in the circuits.
```

File: qiskit/utils/run_circuits.py
Comment (ElePT) [inline qiskit/utils/run_circuits.py:123]:
```suggestion
    An execution wrapper with Qiskit-Terra, with job auto recover capability.
```

File: qiskit/opflow/__init__.py
Comment (Eric-Arellano) [reply qiskit/opflow/__init__.py:25]: This only causes the deprecation to show up when reading docs. It does not issue a runtime warning. For that, you need to call `warnings.warn()` because I did not yet add a helper function for deprecating a module.

File: qiskit/opflow/converters/abelian_grouper.py
Comment (Eric-Arellano) [reply qiskit/opflow/converters/abelian_grouper.py:47]: > Not sure about the "deprecation effort" part of the message.

Yeah, seems fine to remove.

> I now realize I should probably also split this message along 2 lines, the string is too long.

+1. I like using implicit string concatenation like this:

```python
additional_msg=(
   "Sentence 1. "
    "More text."
),
```

Note the space after the period in the first line.

It will not work to use `"""` because it messes up the `add_deprecation_to_docstring` helper function we call via these decorators.

**Comment by woodsp-ibm** [comment]
> Correct, we do not dynamically edit the first sentence to have the one line Deprecated.
> So you may want to revert....

Since these decorators move towards standardizing the user experience I wonder whether us including them manually will just end up with inconsistent docs, where some parts of the docs have this and some do not depending on who does it, as has probably happened until now of course. I think at this point it might be better to leave it to whatever the decorators do? Although I have to admit I like seeing the deprecation status in the indexes that way.

**Comment by Eric-Arellano** [comment]
> Since these decorators move towards standardizing the user experience I wonder whether us including them manually will just end up with inconsistent docs, where some parts of the docs have this and some do not depending on who does it, as has probably happened until now of course. I think at this point it might be better to leave it to whatever the decorators do? Although I have to admit I like seeing the deprecation status in the indexes that way.

I opened https://github.com/Qiskit/qiskit-terra/issues/9876 for that feature request. I think it probably makes sense to have the decorators to do. But as explained there, it's blocked by https://github.com/Qiskit/qiskit-terra/issues/9699. I'm unlikely to have time to implement both those issues in the next two weeks, at least. 

So, I recommend that we revert https://github.com/Qiskit/qiskit-terra/commit/50e5954ecd4a729e7101e8ea4bf493240ae8b59e in the meantime.

File: qiskit/opflow/converters/abelian_grouper.py
Comment (woodsp-ibm) [reply qiskit/opflow/converters/abelian_grouper.py:47]: That whole first sentence can likely go. And since it seems whatever text we end up with is in all of the warnings I do not know if there abetter way to facilitate this.

Comment (Eric-Arellano) [reply qiskit/opflow/converters/abelian_grouper.py:47]: You could always create a private helper function that wraps `@deprecate_func` to DRY things. Emphasis on private because we don't want this to be a long lasting public API; it's only to DRY this specific set of deprecations. 

I'm not sure if that messes up the stacktrace level though -- you would want to experiment with that to confirm.

File: qiskit/opflow/converters/circuit_sampler.py
Comment (woodsp-ibm) [inline qiskit/opflow/converters/circuit_sampler.py:119]: Similarly here.

File: qiskit/opflow/evolutions/matrix_evolution.py
Comment (woodsp-ibm) [inline qiskit/opflow/evolutions/matrix_evolution.py:39]: Maybe its mypy rather than lint which tells one this is a rather useless init. I get that its done for the decorator - if it fails CI I guess we just suppress the error.

File: qiskit/opflow/expectations/expectation_factory.py
Comment (woodsp-ibm) [inline qiskit/opflow/expectations/expectation_factory.py:102]: Again catching warnings...

File: qiskit/utils/measurement_error_mitigation.py
Comment (woodsp-ibm) [inline qiskit/utils/measurement_error_mitigation.py:87]: Is yet to be changed? (Same with other occurrences here)

File: qiskit/opflow/__init__.py
Comment (ElePT) [reply qiskit/opflow/__init__.py:25]: Right, that is why I kept the warning in this page in particular. Once you go to the class docs, you should get the specific deprecation warning for that class.

File: qiskit/opflow/converters/circuit_sampler.py
Comment (ElePT) [reply qiskit/opflow/converters/circuit_sampler.py:86]: Yeah... these are the old changes, there were so many tests that I wanted to let the CI run and see what fails before changing anything.

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:311]: Most of the non-algorithm tests, like this one, use opflow but don't really need it, we could just create the circuit from scratch.

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:698]: After #9537, `QAOAAnstaz` will accept `BaseOperator` as the `cost_operator` type, and opflow will not be necessary for this test.

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:718]: Could use `SparsePauliOp`

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:745]: Could use `SparsePauliOp`

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:772]: Could use `SparsePauliOp`. Same for other instances in this test file.

File: test/python/circuit/library/test_evolution_gate.py
Comment (ElePT) [inline test/python/circuit/library/test_evolution_gate.py:46]: We can rewrite this test file not to use opflow.

File: test/python/transpiler/test_swap_strategy_router.py
Comment (ElePT) [inline test/python/transpiler/test_swap_strategy_router.py:86]: Once https://github.com/Qiskit/qiskit-terra/pull/9537 is merged, we can update this test not to use opflow.

**Review by ElePT — COMMENTED** [review]
I have documented a few non-opflow tests where we could remove opflow dependencies. I am not sure if this PR is the place to do it, or this is the right time to do it (maybe this should be part of the opflow removal PR), but I thought it would be good to know.

**Comment by woodsp-ibm** [comment]
> I have documented a few non-opflow tests where we could remove opflow dependencies. I am not sure if this PR is the place to do it, or this is the right time to do it (maybe this should be part of the opflow removal PR), but I thought it would be good to know.

I would have thought that with opflow emitting deprecation messages that tests using opflow, unless they filtered the warning, already would fail. Hence I would have imagined this PR having to take care of this some way or another to pass CI. Arguably if the use of opflow is optional by that component under test then continuing to test and using opflow until its removed is the right way to go. If its just using opflow as some convenience to build out something in the unit test it would be better to switch that now if we can rather than needing to avert the warning and change it later.

**Review by ElePT — COMMENTED** [review]
A few notes on the latest changes:
1. I have decided to remove the `max_credits` input parameter from `quantum_instance.py`, as it has been deprecated since v.0.20, and having two deprecation decorators at the same time in one class was breaking the code. I think I should include a separate reno documenting this removal, right?
2. I have caught an internal use of opflow in `qiskit/circuit/library/n_local/qaoa_ansatz.py` and replaced it with a quantum info alternative. To check that the class is behaving as expected, I have refactored `test_qaoa_ansatz.py` to use quantum info in all tests but one (which can be removed in the opflow removal PR).
3. I have realized that doing `from opflow import X,Y,Z` and inputting these somewhere doesn't raise any deprecation warning, as the class constructor is not really being called anywhere. On one hand, this is clearly and edge case, but on the other hand I feel like these operators are some of opflow's most heavily used classes. I am open to suggestions on how to fix this. 
4. Following up on Steve's last comment, I have completely removed opflow from all non-opflow-related tests, such as `test/python/transpiler/test_swap_strategy_router.py` and `test/python/circuit/test_circuit_load_from_qpy.py`. I have left the opflow inputs on the evolved operator ansatz, qaoa ansatz and evolution gate because I have considered "opflow-related".

**Comment by woodsp-ibm** [comment]
For max_credits there is this PR #9322 that was opened a while back and updated QI as well

>  have realized that doing from opflow import X,Y,Z and inputting these somewhere doesn't raise any deprecation warning, 

You could try adding a deprecation warning in the opflow \_\_init\_\_.py file so that when that gets loaded it raises a warning, see how that turns out.

File: qiskit/opflow/__init__.py
Comment (woodsp-ibm) [inline qiskit/opflow/__init__.py:22]:
```suggestion
    The :mod:`qiskit.opflow` module is deprecated and will be removed no earlier
```

Comment (woodsp-ibm) [reply qiskit/opflow/__init__.py]: I pushed a change for this.

**Review by woodsp-ibm — APPROVED** [review]
Seems we are finally there with this. It all looks good to me with the code and the API ref docs from it. Thanks @ElePT for picking thus up and seeing it through to the end, and thanks to @manoelmarques for the initial effort this was based on.

File: qiskit/opflow/operator_globals.py
Comment (woodsp-ibm) [reply qiskit/opflow/operator_globals.py:41]: This method is not used outside or operator globals. It could have been private rather than seeming to be public. Either way the intent was that this function was more an internal function in opflow and it is not linked in the API ref. so the spelling without a `d` at the end will only ever be seen if you look at the source code - hence it could just as well be left as it.

**Comment by dongreenberg** [comment]
Oh man can't believe it lasted this long! I too am glad to see the quantum instance behind us, and glad the quantum_info is now far enough along that we don't need duplicate ops and states for the algos. Amazing work on the migration guides! Crazy comprehensive.

**Pull request merged**
